### PR TITLE
[TASK-28] Refresh token

### PR DESCRIPTION
**Pull request by rafayet-monon**
Task Card - https://github.com/rafayet-monon/surveyor/projects/1#card-49663064

## What happened
Redirect to login page if the token is invalid and move login API calls to adapters.
 
## Insight

- [x] Refresh token after logged out
- [x] Move auth API calls to adapters 

## Proof Of Work

- Setting an invalid token and refreshing the page to check that it redirects to login

Untitled_ Dec 15, 2020 2_47 PM

- Tests should pass.

### CODE REVIEW

**Review by carryall — COMMENTED** [review]
@rafayet-monon As mentioned in the requirement 
```
Implement the OAuth authentication including the storage of access and refresh tokens but also the automatic usage of refresh tokens to keep the user logged in.
```

Should we also keep the refresh token from the login API response and use it to call the refresh token API to keep the user logged in after the access token expired 🤔 ?

**Comment by rafayet-monon** [comment]
> Should we also keep the refresh token from the login API response and use it to call the refresh token API to keep the user logged in after the access token expired 🤔 ?

@carryall you are right. The APIs I had was old and didn't have the `refresh_token` API 🙈 
Updated the PR with appropriate changes.

**Comment by rafayetn** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>
<b>🃏 FAIL</b> in <a href="https://github.com/rafayet-monon/surveyor/blob/feature/task-28-refresh-token/src/tests/utils/unauthorizedInterceptor.test.js">unauthorizedInterceptor.test.js</a>


</td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against b57685dd9c9481adf8c01f928cc3416a36a2d8cb
</p>

File: cypress/integration/expiredToken.spec.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+describe('Expired Login', () => {
+  autoRecord();
+
+  beforeEach(() => {
+    cy.login()
+  })
+
+  it('refreshes auth token if token is invalid', function () {
+    cy.get('div[class="home-header__logo"]').find('img').should('be.visible');
```
Comment (carryall) [inline cypress/integration/expiredToken.spec.js]: @rafayet-monon Do we need this line? I think it's not quite related to the topic 🤔

File: src/tests/services/refreshToken.test.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+describe('when refresh token is performed', () => {
+  beforeEach(() => {
+    localStorage.clear();
+  });
+
+  delete window.location;
+  window.location = { reload: jest.fn() };
+
+  it('renders the login page if refresh token is not successful', async () => {
```
Comment (carryall) [inline src/tests/services/refreshToken.test.js]: @rafayet-monon How about
```suggestion
  it('renders the login page if refresh token is missing', async () => {
```

File: src/tests/services/refreshToken.test.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      Promise.resolve({ data: null, status: 400 })
+    );
+
+    RefreshToken();
+
+    expect(localStorage.__STORE__['authorization_token']).toBe(undefined);
+    expect(localStorage.__STORE__['refresh_token']).toBe(undefined);
+  });
+
+  it('updates storage with new token if refresh token is successful', async () => {
```
Comment (carryall) [inline src/tests/services/refreshToken.test.js]:
```suggestion
  it('updates the storage with new tokens if refresh token is valid', async () => {
```

File: src/tests/services/refreshToken.test.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    const refreshToken = RefreshTokenResponse.data.attributes.refresh_token;
+    const accessToken = RefreshTokenResponse.data.attributes.access_token;
+    mockAxios.post.mockImplementation(() =>
+      Promise.resolve({ data: RefreshTokenResponse, status: 200 })
+    );
+
+    RefreshToken();
+
+    expect(localStorage.setItem).not.toHaveBeenLastCalledWith(
```
Comment (carryall) [inline src/tests/services/refreshToken.test.js]: @rafayet-monon Not sure about this, but should we use `expect(localStorage.setItem).toHaveBeenLastCalledWith` to ensure the token is set?

**Comment by carryall** [comment]
@rafayet-monon I'm not sure that we should cover this case when the `authorization_token` value is missing in the storage or not 🤔 
react_refresh_token

**Comment by rafayet-monon** [comment]
@carryall Fixed test related reviews in - [a988c94](https://github.com/rafayet-monon/surveyor/pull/23/commits/a988c94f9b187bffee59fad1176fa502018b5700) and refresh token issue when auth token is missing fixed in. - [fad6dcd](https://github.com/rafayet-monon/surveyor/pull/23/commits/fad6dcd23fb225a4055474d16ab26fd0e69658c5)

**Review by Lahphim — APPROVED** [review]

File: cypress/integration/expiredToken.spec.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+const autoRecord = require('cypress-autorecord');
+
+describe('Expired Login', () => {
+  autoRecord();
+
+  beforeEach(() => {
+    cy.login()
+  })
+
+  it('refreshes auth token if token is invalid', function () {
```
Comment (carryall) [inline cypress/integration/expiredToken.spec.js]: @rafayet-monon one small comment here, would it be better if we use an arrow function here? what do you think?
```suggestion
  it('refreshes auth token if token is invalid', () => {
```

**Review by carryall — APPROVED** [review]

**Pull request merged**